src/hotspot/share/gc/shared/cardTable.cpp

```diff
@@ -92,10 +89,12 @@ CardTable::~CardTable() {
 }
 
 void CardTable::initialize() {
-  _guard_index = cards_required(_whole_heap.word_size()) - 1;
-  _last_valid_index = _guard_index - 1;
+  size_t num_cards = cards_required(_whole_heap.word_size());
+  _last_valid_index = num_cards - 1;
 
-  _byte_map_size = compute_byte_map_size();
+  // each card takes 1 byte; + 1 for the guard card
+  size_t num_bytes = num_cards + 1;
+  _byte_map_size = compute_byte_map_size(num_bytes);
```

Comment:
This seems a pre-existing bug/oddity, but `_byte_map_size` includes the guard card; there are several asserts (in `CardTable::byte_for` and `PSCardTable::is_valid_card_address`) that check:
```
assert(result >= _byte_map && result < _byte_map + _byte_map_size,
```
which seems weird to me as the card at the guard card is somehow assumed to be a valid part of the card table.
I believe the `_byte_map_size` should not include the guard card itself (which we replaced by a guard page where access to it will cause an immediate error anyway), but pass the `+1` to the `ReservedSpace` constructor directly (and maybe adjust other locations that use it to print stuff, i.e. introduce a helper method that adds that +1?)

What do you think?

Reply:
That assert is indeed too weak. Maybe the longer-term plan can be to drop guard card/page completely, like what's in G1. Having "random" `+1` in the code could be confusing.